blobstore/src/main/java/org/jclouds/blobstore/util/BlobStoreUtils.java

```diff
@@ -85,4 +103,65 @@ private static void convertUserMetadataKeysToLowercase(MutableBlobMetadata metad
       }
       metadata.setUserMetadata(lowerCaseUserMetadata);
    }
+
+   private static class CrawlBlobStoreIterable implements Iterable<StorageMetadata> {
+      private final BlobStore blobStore;
+      private final String containerName;
+      private final ListContainerOptions options;
+
+      CrawlBlobStoreIterable(BlobStore blobStore, String containerName, ListContainerOptions options) {
+         this.blobStore = checkNotNull(blobStore);
+         this.containerName = checkNotNull(containerName);
+         this.options = checkNotNull(options);
+      }
+
+      @Override
+      public Iterator<StorageMetadata> iterator() {
+         return new CrawlBlobStoreIterator(blobStore, containerName, options);
+      }
+   }
+
+   private static class CrawlBlobStoreIterator implements Iterator<StorageMetadata> {
+      private final BlobStore blobStore;
+      private final String containerName;
+      private final ListContainerOptions options;
+      private Iterator<? extends StorageMetadata> iterator;
+      private String marker;
+
+      CrawlBlobStoreIterator(BlobStore blobStore, String containerName, ListContainerOptions options) {
+         this.blobStore = checkNotNull(blobStore);
+         this.containerName = checkNotNull(containerName);
+         this.options = checkNotNull(options).recursive();
+         advance();
+      }
+
+      private void advance() {
+         if (marker != null) {
+            options.afterMarker(marker);
+         }
+         PageSet<? extends StorageMetadata> set = blobStore.list(containerName, options);
+         marker = set.getNextMarker();
+         iterator = set.iterator();
+      }
+
+      @Override
+      public boolean hasNext() {
+         return iterator.hasNext() || marker != null;
+      }
+
+      @Override
+      public StorageMetadata next() {
+         while (true) {
+            if (!iterator.hasNext()) {
+               advance();
+            }
+            StorageMetadata metadata = iterator.next();
```

Comment:
Looking at your implementation, that shouldn't be needed. That concat would substitute your `while(true)` loop.

In the links I passed as an example, you have an `Iterator<Page<T>>`, that is, an iterator that knows how to "fetch pages", but does not iterate the elements of each page. The `concat` method is used to "flatten" that iterator in an iterator that iterates over the element of each page and jumps to the next one once finished. Basically, if you model it to be an Iterator of pages (which is what it really is), you can take benefit of Guava's `Iterators.concat(Iterator)` to do the "jump from one page to the next one" for you.

It is just an implementation detail. I just shared how pagination is implemented in other parts of the code (which IMO ends up being simpler code). The current implementation looks fine, BTW :)